packages/muelu/src/Transfers/Smoothed-Aggregation/MueLu_SaPFactory_def.hpp

```diff
@@ -247,6 +257,103 @@ namespace MueLu {
 
   } //Build()
 
+  // Analyze the grid transfer produced by smoothed aggregation and make
+  // modifications if it does not look right. In particular, if there are
+  // negative entries or entries larger than 1, modify P's rows. 
+  //
+  // Note: this kind of evaluation probably only makes sense if not doing QR
+  // when constructing tentative P.
+  //
+  // For entries that do not satisfy the two constraints (>= 0  or <=1) we set
+  // these entries to the constraint value and modify the rest of the row
+  // so that the row sum remains the same as before by adding an equal
+  // amount to each remaining entry. However, if the original row sum value
+  // violates the constraints, we set the row sum back to 1 (the row sum of 
+  // tentative P). After doing the modification to a row, we need to check 
+  // again the entire row to make sure that the modified row does not violate
+  // the constraints.
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void SaPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::SatisfyPConstraints(const RCP<Matrix> A, RCP<Matrix>& P) const {
+
+    LO nPDEs = A->GetFixedBlockSize();
+    Teuchos::ArrayRCP<Scalar> ConstraintViolationSum(nPDEs);
+    Teuchos::ArrayRCP<Scalar> Rsum(nPDEs);
+    Teuchos::ArrayRCP<size_t> nPositive(nPDEs);
+    for (size_t k=0; k < (size_t) nPDEs; k++) ConstraintViolationSum[k] = Teuchos::ScalarTraits<Scalar>::zero(); 
+    for (size_t k=0; k < (size_t) nPDEs; k++) Rsum[k] = Teuchos::ScalarTraits<Scalar>::zero(); 
+    for (size_t k=0; k < (size_t) nPDEs; k++) nPositive[k] = 0;
+
+
+    for (size_t i = 0; i < as<size_t>(P->getRowMap()->getNodeNumElements()); i++) {
+
+      Teuchos::ArrayView<const LocalOrdinal> indices;
+      Teuchos::ArrayView<const Scalar> vals1;
+      Teuchos::ArrayView<      Scalar> vals;
+      P->getLocalRowView((LO) i, indices, vals1);
+      size_t nnz = indices.size();
+      if (nnz == 0) continue;
+
+      vals = ArrayView<Scalar>(const_cast<SC*>(vals1.getRawPtr()), nnz);
+
+
+      bool checkRow = true;
+
+      while (checkRow) { 
+
+        // check constraints and compute the row sum
+    
+        for (LO j = 0; j < indices.size(); j++)  {
+          Rsum[ j%nPDEs ] += vals[j]; 
+          if (Teuchos::ScalarTraits<SC>::real(vals[j]) < Teuchos::ScalarTraits<SC>::real(Teuchos::ScalarTraits<Scalar>::zero())) { 
```

Comment:
You could typedef `Teuchos::ScalarTraits<Scalar>::zero()` or do `Scalar zero = Teuchos::ScalarTraits<Scalar>::zero()`, which would make the code easier to read.

Reply:
@jhux2  Added the Scalar zero = ... stuff so now the code is cleaner.